### ISSUE TITLE
On new VM install our software development environment with third-party libraries

### ISSUE THREAD

**Issue by Ross-ROE** [wfau/ScienceArchives#559]

**Comment by Ross-ROE** [comment]
We agreed in today's meeting to install Python3 via OS's package manager, and then install Python libraries via pip at the system level for all users. Though we may need sudo access to do this.

**Comment by Ross-ROE** [comment]
I just checked and there is already a system Python3 installed: though it's older, Python 3.10.12.

It would be best to somehow install our own Python3 environment with a more up-to-date version: Python 3.12.4 is the latest. Not sure what the best way to install two separate Python3 environments in parallel is - would need to be sure that pip is installing the libraries for the right Python.

Checking online it seems that for our Ubuntu system it's best to use the "dead snakes" repo (sounds dodgy, but it's just the Python reference I guess):
`$ sudo apt-get install software-properties-common
$ sudo add-apt-repository ppa:deadsnakes/ppa
$ sudo apt-get update
$ sudo apt-get install python3.12`

**Comment by astrodb** [comment]
I suggest building the Pytho environment with Anaconda. That way it'll be independent of the OS, can be easily customized by users, and easily exported to other machines when needed.

**Comment by zexpe** [comment]
The concern we had with Anaconda was we wanted to use pip to manage the libraries. Would that be OK with Anaconda - for a system-wide common library install? Also what advantage would Anaconda have over the apt-get install from the dead snakes repo? It's independent of the OS, can be customised and installed elsewhere. Is the concern it's tied to Ubuntu still - so if want to change Linux distro it would be a hassle?

**Comment by dr-bob** [comment]
> I suggest building the Pytho environment with Anaconda. That way it'll be independent of the OS, can be easily customized by users, and easily exported to other machines when needed.

Do we want it to be "easily customized by users"?

**Comment by zexpe** [comment]
No. It should be fairly static and managed via request really.

**Comment by astrodb** [comment]
Using pip to install with Anaconda is fine, so long as you direct pip accordingly. 

apt-get is not independent from the OS install, it is the OS install. Using that is fine, but it isn't as portable or flexible. It means you're stuck to Ubuntu (which is fine), it gets updated automatically with the OS (which sometimes can cause problems), and it means you have to should be more careful about taking snapshots of the VM before making changes to the environment to ensure you have a working backup.

By using Anaconda you can build your env separate from the OS, update it separate from the OS, and quickly switch between DEV and PROD environments. It also means you can copy the env to another VM with ease and backup existing env's with ease.

**Comment by zexpe** [comment]
Presumably a Python install from the dead snakes repo wouldn't automatically update with the OS?

I don't mind - we prefer to use pip for library management as we all know it and trust it well, and it's the most standard way for any Python library to be published. So as long as we can have a common environment managed by pip we should all be happy - wherever the Python environment comes from.

**Comment by dr-bob** [comment]
> My comment about "easily customized by users" meant user accounts on the VM, not actual users! (I meant our devs)

I know. Do we want it to be easily customised by our devs? :-S

**Comment by astrodb** [comment]
At the start, allowing the devs to change the env is a good idea from my perspective as it saves me the trouble of installing packages. Later on it makes sense to keep the changes minimal, but that can be done with either solution.

My hesitation with dead snakes repo is that they provide no guarantees of security updates, and yet it installs as the default system Python env. This means we'd have to pay closer attention to Python updates and patch things ourselves when holes are found.

Anaconda keeps the OS Python independent from the Archive Python, thus reducing the security footprint.

**Comment by astrodb** [comment]
Note the opening comment on security here: https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa

**Comment by dr-bob** [comment]
On a slightly different note, what are the login details for the Somerville server?

**Comment by astrodb** [comment]
> On a slightly different note, what are the login details for the Somerville server?

The login for Somerville is managed with IRIS IAM, allowing control of the project (essentially the Sys admin tasks).
The logins for the moons-archive VM are just standard Linux accounts I created based on the details of our existing servers at ROE. Let me know if you want an account and I'll create it (and send me your SSH public key).

**Comment by zexpe** [comment]
I didn't think dead snakes replaces system Python - I thought it was supposed to be a separate copy? Anyway, if you have no problem installing Anaconda and we go with a pip managed library - let's go for that.

**Comment by astrodb** [comment]
Looking closer, deadsnakes does go to a separate location, but it's still exposed via the OS. So not as insecure as I thought, but still a minor concern.

Conda can be installed directly by the users, no need for root. From there you want to follow instructions in the first answer here to install the env, select it, install pip, and use pip: https://stackoverflow.com/questions/41060382/using-pip-to-install-packages-to-anaconda-environment

Ultimately I'd recommend creating a 'dev' enviroment in a known location, build it until it works, then copy it to a directory named 'prod' for using in the production system. Then go on updating the 'dev' env as needed, taking backups of the 'prod' env each time before copying in the new 'dev' env when they're deemed ready.

**Comment by Ross-ROE** [comment]
Sounds overly complicated for our purposes! Really, I just want something like EDEN that has everything I need & I don't need to worry about it... if we need to update library versions and it breaks something that's not really a problem, we just revert to the previous version by forcing that version in pip. No need to complicate things with multiple environments. The main thing is that every is running the code against a common environment.

**Comment by esutorius** [comment]
I'm full with you there, Ross. I'd say we install an Anaconda-Python under /home/scos/anaconda3 and any additional non-python software we might need at some point (and that we don't want as a general root installation) in /home/scos/usr/local 
This would roughly match what we have now and what we're used to.

**Comment by Ross-ROE** [comment]
That's fine... though I thought we'd also decided not to copy the "scos" user model? So we can have the environment installed anywhere really.

**Comment by esutorius** [comment]
My last comment seems to have vanished... I meant /home/etws , I just had the wrong hat on. ;)

I've installed anaconda and psycopg (with pip, no problem there), but we might need to setup multiusers as specified here:
https://docs.anaconda.com/anaconda/install/multi-user/

I've also fixed the annoying "libtinfo.so.6: no version information available" warning messages by re-installing ncurses from the conda-forge.

To check if you can use the installation, use the PATH addition from /home/etws/.login (don't know if you'd need LD_*_PATHs) and the last line from /home/etws/.tcshrc so anaconda starts properly after login. (If you're using bash, use the /home/etws/.bashrc file instead.) 
If you want to have the 'base' environment available after you log in set
`# The base environment is activated by default
conda config --set auto_activate_base True
`
Otherwise set it to False.

**Comment by Ross-ROE** [comment]
Thanks Eckhard. I'm able to use your Python environment by just adding to my PATH:
`
moons-archive:/home/rsc> setenv PATH /home/etws/anaconda3/bin:${PATH}
moons-archive:/home/rsc> python3
Python 3.12.3 | packaged by conda-forge | (main, Apr 15 2024, 18:38:13) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import psycopg
>>> psycopg.__version__
'3.2.1'
>>> import astropy
>>> astropy.__version__
'6.1.0'
`

So it looks like I'm good to go with development from that point of view. If I want an extra library installed - shall I just request for you to install it?

**Comment by dr-bob** [comment]
Before anyone does anything in depth with this, I can't express strongly enough how colossally bad it is to have this environment set up in somebody's home directory. We should really set it up in a centralised, system-wise location and have everyone use the same installation.

**Comment by esutorius** [comment]
Yes, sounds sensible, but nobody did it and as normal I have no root/sudo rights, so the only way to progress this is to do it the 'scos' way with a different user name. ;)

**Comment by Ross-ROE** [comment]
Is that the conda version of Astropy? It appears to be two releases behind the latest.

**Comment by esutorius** [comment]
No problem, I've upgraded it now via pip, but it's still only 6.1.2 so only a revision change.

**Comment by Ross-ROE** [comment]
Yea, that was one of my concerns with conda - you get a random bunch of versions that they've packaged up together. With things like Astropy I'm likely to be reporting bugs and want to be able to get the latest version without waiting on conda. pip also has literally everything.

**Comment by Ross-ROE** [comment]
Likewise, I see the latest version of Astropy is tested against Numpy 2.0, whilst conda has installed 1.26.4. 🤦‍♂️

**Comment by Ross-ROE** [comment]
I'd feel safer having no libraries, and let pip manage all the dependencies than have a bunch of outdated conda libraries pre-installed.

**Comment by dr-bob** [comment]
Then maybe miniconda is the way to go:
https://docs.anaconda.com/miniconda/

"It is a small bootstrap version of Anaconda that includes only conda, Python, the packages they both depend on, and a small number of other useful packages (like pip, zlib, and a few others)."

Then install everything else via pip?

**Comment by esutorius** [comment]
Numpy is now 2.0.1, which broke dependencies on contourpy, streamlit, pywavelets, numba, which I've updated now as well.
I'm also trying to run a conda update all using the conda-forge. Hopefully that'll do the trick.

**Comment by Ross-ROE** [comment]
I'd just start from scratch using an empty install like Rob suggests. Get pip to install Astropy and Psycopg3 - and all the dependencies it needs. Avoid the chaos of the conda install.

**Comment by esutorius** [comment]
The only problem I see there is that every user ends with their own installation which can be inconsistent unless you run a daily update.

**Comment by Ross-ROE** [comment]
We need a common environment - but that can be managed entirely by pip - we don't need conda for a common environment.

**Comment by dr-bob** [comment]
I'll set up a centralised environment and see if I can make it so everyone can update with pip. If not, I'll give everyone sudo privileges to run pip.

**Comment by Ross-ROE** [comment]
So by default pip installs in the site-packages directory of the Python installation, unless you specify to install it locally in your user directory. The access rights to that site-packages directory can be set to anything... doesn't have to be root, but we also don't need to give all users access to write there. Having one user manage the install is probably safest. Also, by having a clean install we know exactly what our dependencies are, we are not using some underlying library that was installed by default without realising it.

**Comment by Ross-ROE** [comment]
Talking of which, once we settled on a final solution, we should document the installation procedure, as well as the overall VM setup here: https://github.com/wfau/ScienceArchives/wiki

**Comment by astrodb** [comment]
I've created a common directory for the python environment in: /moons-archive/python

I've also created a page with the server details here: https://github.com/wfau/ScienceArchives/wiki/Server-Configuration-Details

**Comment by Ross-ROE** [comment]
Thanks for the documentation! I also had that thought about using the /moons-archive directory for the software install... but then thought - what if we want to curate multiple archives on the same machine? We'd want to use common software for that... so probably best somewhere else - keep /moons-archive for just the database and products?

**Comment by astrodb** [comment]
OK, I've deleted the python directory in /moons-archive, and created a shared directory called: /home/dev-env

**Comment by Ross-ROE** [comment]
> > My comment about "easily customized by users" meant user accounts on the VM, not actual users! (I meant our devs)
> 
> I know. Do we want it to be easily customised by our devs? :-S

It would be nice if anyone could easily add libraries or update libraries, but we would still need to have some kind of charge control process, just to let people know what's changed - and if there's a problem we can then easily revert. This could just be as simple as updating a wiki page to say what you did and what pip changed. Or we could have an open issue like this just to monitor config change?

**Comment by esutorius** [comment]
The latter sounds like a good idea.  Whoever does that and has access to dev-env and installs things, I've downloaded the miniconda installer into /home/etws/downloads/  It just needs copying over and run with 'bash Miniconda...'

**Comment by Ross-ROE** [comment]
I think we've missed a trick here, with the Euclid Dev Env called EDEN, we could have called ours SADEN.

**Comment by dr-bob** [comment]
I've installed miniconda into /home/dev-env/miniconda3. Everything is owned by root just now, so likely nobody will be able to update anything. I'm not sure what's the best way to figure out which directories need their permissions changing. Will it just be /home/dev-env/miniconda3/lib/python3.12/site-packages?

Or maybe someone can just try installing something with pip and let me know what breaks.

**Comment by Ross-ROE** [comment]
> /home/dev-env/miniconda3/lib/python3.12/site-packages

Yes at a minimum that directory would need to be writable:
<pre>
moons-archive:~> ls -ld /home/dev-env/miniconda3/lib/python3.12/site-packages
drwxr-xr-x 73 root root 4096 Aug  1 13:18 /home/dev-env/miniconda3/lib/python3.12/site-packages
</pre>

**Comment by dr-bob** [comment]
OK, that's writeable now. Let me know if you need anything else fixing.

**Comment by Ross-ROE** [comment]
Hmmm...
<pre>
moons-archive:~> pip3 install astropy
Defaulting to user installation because normal site-packages is not writeable
</pre>

**Comment by Ross-ROE** [comment]
I should have noted too:
<pre>
moons-archive:~> which pip3
/home/dev-env/miniconda3/bin/pip3
</pre>
There must be another environment variable I need to change?

**Comment by Ross-ROE** [comment]
This looks fine too:
<pre>
moons-archive:~> python3
Python 3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 15:12:24) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/home/dev-env/miniconda3/lib/python312.zip', '/home/dev-env/miniconda3/lib/python3.12', '/home/dev-env/miniconda3/lib/python3.12/lib-dynload', '/home/rsc/.local/lib/python3.12/site-packages', '/home/dev-env/miniconda3/lib/python3.12/site-packages']
</pre>

**Comment by Ross-ROE** [comment]
Manually specifying the target appeared to work and then barfed:
<pre>
moons-archive:~> pip3 install --target=/home/dev-env/miniconda3/lib/python3.12/site-packages astropy
Collecting astropy
  Using cached astropy-6.1.2-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (10 kB)
Collecting numpy>=1.23 (from astropy)
  Using cached numpy-2.0.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (60 kB)
Collecting pyerfa>=2.0.1.1 (from astropy)
  Using cached pyerfa-2.0.1.4-cp39-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.7 kB)
Collecting astropy-iers-data>=0.2024.7.1.0.34.3 (from astropy)
  Using cached astropy_iers_data-0.2024.7.29.0.32.7-py3-none-any.whl.metadata (5.1 kB)
Collecting PyYAML>=3.13 (from astropy)
  Using cached PyYAML-6.0.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.1 kB)
Collecting packaging>=19.0 (from astropy)
  Downloading packaging-24.1-py3-none-any.whl.metadata (3.2 kB)
Using cached astropy-6.1.2-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.3 MB)
Using cached astropy_iers_data-0.2024.7.29.0.32.7-py3-none-any.whl (1.9 MB)
Using cached numpy-2.0.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (19.2 MB)
Downloading packaging-24.1-py3-none-any.whl (53 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.0/54.0 kB 1.6 MB/s eta 0:00:00
Using cached pyerfa-2.0.1.4-cp39-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (738 kB)
Using cached PyYAML-6.0.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (724 kB)
Installing collected packages: PyYAML, packaging, numpy, astropy-iers-data, pyerfa, astropy
Successfully installed PyYAML-6.0.1 astropy-6.1.2 astropy-iers-data-0.2024.7.29.0.32.7 numpy-2.0.1 packaging-24.1 pyerfa-2.0.1.4
ERROR: Exception:
Traceback (most recent call last):
  File "/home/dev-env/miniconda3/lib/python3.12/shutil.py", line 847, in move
    os.rename(src, real_dst)
PermissionError: [Errno 13] Permission denied: '/tmp/pip-target-eguc_p5a/lib/python/astropy_iers_data-0.2024.7.29.0.32.7.dist-info' -> '/home/dev-env/miniconda3/lib/python3.12/site-packages/astropy_iers_data-0.2024.7.29.0.32.7.dist-info'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dev-env/miniconda3/lib/python3.12/site-packages/pip/_internal/cli/base_command.py", line 180, in exc_logging_wrapper
    status = run_func(*args)
             ^^^^^^^^^^^^^^^
  File "/home/dev-env/miniconda3/lib/python3.12/site-packages/pip/_internal/cli/req_command.py", line 245, in wrapper
    return func(self, options, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev-env/miniconda3/lib/python3.12/site-packages/pip/_internal/commands/install.py", line 510, in run
    self._handle_target_dir(
  File "/home/dev-env/miniconda3/lib/python3.12/site-packages/pip/_internal/commands/install.py", line 569, in _handle_target_dir
    shutil.move(os.path.join(lib_dir, item), target_item_dir)
  File "/home/dev-env/miniconda3/lib/python3.12/shutil.py", line 863, in move
    copytree(src, real_dst, copy_function=copy_function,
  File "/home/dev-env/miniconda3/lib/python3.12/shutil.py", line 600, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev-env/miniconda3/lib/python3.12/shutil.py", line 498, in _copytree
    os.makedirs(dst, exist_ok=dirs_exist_ok)
  File "<frozen os>", line 225, in makedirs
PermissionError: [Errno 13] Permission denied: '/home/dev-env/miniconda3/lib/python3.12/site-packages/astropy_iers_data-0.2024.7.29.0.32.7.dist-info'
</pre>

**Comment by dr-bob** [comment]
That's odd. What happens if you run the following?

cd /home/dev-env/miniconda3/lib/python3.1/site-packages
mkdir astropy_iers_data-0.2024.7.29.0.32.7.dist-info

**Comment by Ross-ROE** [comment]
mkdir: cannot create directory ‘astropy_iers_data-0.2024.7.29.0.32.7.dist-info’: Permission denied

**Comment by esutorius** [comment]
Since that is a file, could it be that a permission is not set on any of the directories before site-packages?

**Comment by dr-bob** [comment]
I think the problem is that I set write permission to the "users" group, but none of you were in that group. So then I added you in, but if you were already logged in it probably hasn't updated your existing session. Try logging out and in again and see what happens.

**Comment by esutorius** [comment]
You could also try the way that the anaconda multi-user environment is set up (see the link above).

**Comment by dr-bob** [comment]
Those instructions are for Anaconda, not miniconda. I followed the instructions for installing miniconda, but it didn't give me any options for installing it in multi-user mode :-S

**Comment by Ross-ROE** [comment]
So logging out and back in, I can now install without specifying a target, but I hit a new permissions error:
<pre>
moons-archive:~> pip3 install astropy
Collecting astropy
  Using cached astropy-6.1.2-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (10 kB)
Requirement already satisfied: numpy>=1.23 in ./.local/lib/python3.12/site-packages (from astropy) (2.0.1)
Requirement already satisfied: pyerfa>=2.0.1.1 in ./.local/lib/python3.12/site-packages (from astropy) (2.0.1.4)
Requirement already satisfied: astropy-iers-data>=0.2024.7.1.0.34.3 in ./.local/lib/python3.12/site-packages (from astropy) (0.2024.7.29.0.32.7)
Requirement already satisfied: PyYAML>=3.13 in ./.local/lib/python3.12/site-packages (from astropy) (6.0.1)
Requirement already satisfied: packaging>=19.0 in /home/dev-env/miniconda3/lib/python3.12/site-packages (from astropy) (23.2)
Using cached astropy-6.1.2-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.3 MB)
Installing collected packages: astropy
ERROR: Could not install packages due to an OSError: [Errno 13] Permission denied: '/home/dev-env/miniconda3/bin/fits2bitmap'
Consider using the `--user` option or check the permissions.
</pre>

**Comment by dr-bob** [comment]
OK, I've made /home/dev-env/miniconda3/bin writeable by the "users" group. 

Try again (again).

**Comment by Ross-ROE** [comment]
Success!
<pre>
moons-archive:~> pip3 install astropy
Collecting astropy
  Using cached astropy-6.1.2-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (10 kB)
Requirement already satisfied: numpy>=1.23 in ./.local/lib/python3.12/site-packages (from astropy) (2.0.1)
Requirement already satisfied: pyerfa>=2.0.1.1 in ./.local/lib/python3.12/site-packages (from astropy) (2.0.1.4)
Requirement already satisfied: astropy-iers-data>=0.2024.7.1.0.34.3 in ./.local/lib/python3.12/site-packages (from astropy) (0.2024.7.29.0.32.7)
Requirement already satisfied: PyYAML>=3.13 in ./.local/lib/python3.12/site-packages (from astropy) (6.0.1)
Requirement already satisfied: packaging>=19.0 in /home/dev-env/miniconda3/lib/python3.12/site-packages (from astropy) (23.2)
Using cached astropy-6.1.2-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.3 MB)
Installing collected packages: astropy
Successfully installed astropy-6.1.2
</pre>

**Comment by dr-bob** [comment]
Weird. There's a subdirectory of site-packages called "~ackaging-23.2.dist-info".

Maybe an install that went wrong, or something. I'm guessing it's probably best if I just delete it.

**Comment by esutorius** [comment]
I think those are leftover temporary directories when something broke during install. 
BTW, where are you getting your code from? Maybe use the pip.conf (see the one in /home/etws/pip/) to only connect to trusted sources.

**Comment by Ross-ROE** [comment]
Yea, that does seem to be the case. Removing those files has removed the error. It's just using whatever the miniconda pip is set to use. So far I've installed, astropy, scipy, and psycopg:
<pre>
Successfully installed PyYAML-6.0.1 astropy-6.1.2 astropy-iers-data-0.2024.7.29.0.32.7 numpy-2.0.1 packaging pyerfa-2.0.1.4
Successfully installed scipy-1.14.0
Successfully installed psycopg-3.2.1 typing-extensions-4.12.2
</pre>

**Comment by tms-epcc** [comment]
04/APR/25
- investigating how to make this more maintainable

**Comment by tms-epcc** [comment]
17/DEC/25
- Ross considering doing this differently for better long term maintenance 
- low priority
